src/vs/code/electron-main/app.ts

```diff
@@ -555,13 +568,277 @@ export class CodeApplication extends Disposable {
+			getActiveWindowId: () => nativeHostMainService.getActiveWindowId(-1)
+		}));
+		const activeWindowRouter = new StaticRouter(ctx => activeWindowManager.getActiveClientId().then(id => ctx === id));
+		const urlHandlerRouter = new URLHandlerRouter(activeWindowRouter);
+		const urlHandlerChannel = mainProcessElectronServer.getChannel('urlHandler', urlHandlerRouter);
+		urlService.registerHandler(new URLHandlerChannelClient(urlHandlerChannel));
+
+		// Watch Electron URLs and forward them to the UrlService
+		this._register(new ElectronURLListener(initialProtocolUrls, urlService, windowsMainService, this.environmentMainService, this.productService, this.logService));
+
+		return initialProtocolUrls;
+	}
+
+	private resolveInitialProtocolUrls(): IInitialProtocolUrls {
+
+		// Windows/Linux: protocol handler invokes CLI with --open-url
+		const protocolUrlsFromCommandLine = this.environmentMainService.args['open-url'] ? this.environmentMainService.args._urls || [] : [];
+		if (protocolUrlsFromCommandLine.length) {
+			this.logService.trace('app#resolveInitialProtocolUrls() protocol urls from command line:', protocolUrlsFromCommandLine);
+		}
+
+		// macOS: open-url events that were received before the app is ready
+		const protocolUrlsFromEvent = ((<any>global).getOpenUrls() || []) as string[];
+		if (protocolUrlsFromEvent.length) {
+			this.logService.trace(`app#resolveInitialProtocolUrls() protocol urls from macOS 'open-url' event:`, protocolUrlsFromEvent);
+		}
+
+		// Go over all protocol urls and split window openables
+		// from other protocol urls that need to be handled within
+		// the app.
+		// Filter out blocked urls or invalid ones.
+
+		const openables: (IWindowOpenable | IOpenEmptyWindowOptions)[] = [];
+		const urls = [
+			...protocolUrlsFromCommandLine,
+			...protocolUrlsFromEvent
+		].map(url => {
+			try {
+				return { uri: URI.parse(url), originalUrl: url };
+			} catch {
+				this.logService.trace('app#resolveInitialProtocolUrls() protocol url failed to parse:', url);
+
+				return undefined;
+			}
+		}).filter((obj): obj is IProtocolUrl => {
+			if (!obj) {
+				return false; // invalid
+			}
+
+			if (this.shouldBlockURI(obj.uri)) {
+				this.logService.trace('app#resolveInitialProtocolUrls() protocol url was blocked:', obj.uri.toString(true));
+
+				return false; // blocked
+			}
+
+			const windowOpenable = this.getWindowOpenableFromProtocolUrl(obj.uri);
+			if (windowOpenable) {
+				this.logService.trace('app#resolveInitialProtocolUrls() protocol url will be handled as window to open:', obj.uri.toString(true), windowOpenable);
+
+				openables.push(windowOpenable);
+
+				return false; // handled as window to open
+			}
+
+			const params = new URLSearchParams(obj.uri.query);
+			if (params.get('windowId') === '_blank') {
+				this.logService.trace(`app#resolveInitialProtocolUrls() found 'windowId=_blank' as parameter, protocol url will be handled as empty window to open:`, obj.uri.toString(true));
```

Comment:
This is pretty much the core of the fix for https://github.com/github/codespaces/issues/9792: treat a protocol URL that needs to be handled by an extension and has `windowId=_blank` as an empty window to open (via `IOpenEmptyWindowOptions`) so that we do not restore previous sessions.